### PR TITLE
[CSV-206] Add support for String delimiters

### PR DESCRIPTION
**Pull request by dota17**
https://issues.apache.org/jira/browse/CSV-206

### CODE REVIEW

**Comment by melloware** [comment]
Looks like code coverage decreased and its failing the build.  You will need to add more unit tests for coverage.

**Comment by dota17** [comment]
ok, I will add more unit tests.

**Comment by dota17** [comment]
Any comments here?

**Comment by san4you88** [comment]
Hi Team,
  Am waiting for the fix from long time, will we be merging this PR
I can speand sometime and resolve the merge conflicts if any1 needs help

File: src/test/java/org/apache/commons/csv/issues/JiraCsv213Test.java [hunk 1]
```diff
@@ -43,7 +43,7 @@ private void createEndChannel(final File csvFile) {
         // @formatter:off
         final CSVFormat csvFormat =
                 CSVFormat.DEFAULT
-                    .withDelimiter(';')
+                    .withDelimiter(";")
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/issues/JiraCsv213Test.java:46]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/issues/JiraCsv211Test.java [hunk 2]
```diff
@@ -32,11 +32,11 @@
     public void testJiraCsv211Format() throws IOException {
         final String[] values = new String[] { "1", "Jane Doe", "USA", "" };
 
-        final CSVFormat printFormat = CSVFormat.DEFAULT.withDelimiter('\t').withHeader("ID", "Name", "Country", "Age");
+        final CSVFormat printFormat = CSVFormat.DEFAULT.withDelimiter("\t").withHeader("ID", "Name", "Country", "Age");
         String formatted = printFormat.format(values);
         assertEquals("ID\tName\tCountry\tAge\r\n1\tJane Doe\tUSA\t", formatted);
 
-        final CSVFormat parseFormat = CSVFormat.DEFAULT.withDelimiter('\t').withFirstRecordAsHeader();
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/issues/JiraCsv211Test.java:39]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/issues/JiraCsv211Test.java [hunk 3]
```diff
@@ -32,11 +32,11 @@
     public void testJiraCsv211Format() throws IOException {
         final String[] values = new String[] { "1", "Jane Doe", "USA", "" };
 
-        final CSVFormat printFormat = CSVFormat.DEFAULT.withDelimiter('\t').withHeader("ID", "Name", "Country", "Age");
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/issues/JiraCsv211Test.java:35]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/issues/JiraCsv167Test.java [hunk 4]
```diff
@@ -55,7 +55,7 @@ public void parse() throws IOException {
         //
         format = format.withAllowMissingColumnNames(false);
         format = format.withCommentMarker('#');
-        format = format.withDelimiter(',');
+        format = format.withDelimiter(",");
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/issues/JiraCsv167Test.java:58]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/LexerTest.java [hunk 5]
```diff
@@ -258,7 +258,7 @@ public void testNextToken6() throws IOException {
          * file: a;'b and \' more ' !comment;;;; ;;
          */
         final String code = "a;'b and '' more\n'\n!comment;;;;\n;;";
-        final CSVFormat format = CSVFormat.DEFAULT.withQuote('\'').withCommentMarker('!').withDelimiter(';');
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/LexerTest.java:261]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVRecordTest.java [hunk 6]
```diff
@@ -257,7 +257,7 @@ public void testToMap() {
 
     @Test
     public void testToMapWithNoHeader() throws Exception {
-        try (final CSVParser parser = CSVParser.parse("a,b", CSVFormat.newFormat(','))) {
+        try (final CSVParser parser = CSVParser.parse("a,b", CSVFormat.newFormat(","))) {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVRecordTest.java:260]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVPrinterTest.java [hunk 7]
```diff
@@ -602,7 +634,7 @@ public void testHeaderNotSet() throws IOException {
 
     @Test
     public void testInvalidFormat() {
-        assertThrows(IllegalArgumentException.class, () -> CSVFormat.DEFAULT.withDelimiter(CR));
+        assertThrows(IllegalArgumentException.class, () -> CSVFormat.DEFAULT.withDelimiter(String.valueOf(CR)));
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVPrinterTest.java:637]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVParserTest.java [hunk 8]
```diff
@@ -1229,7 +1229,7 @@ private void validateRecordPosition(final String lineSeparator) throws IOExcepti
             // completeness...
             "\u00c4,\u00d6,\u00dc" + lineSeparator + "EOF,EOF,EOF";
 
-        final CSVFormat format = CSVFormat.newFormat(',').withQuote('\'').withRecordSeparator(lineSeparator);
+        final CSVFormat format = CSVFormat.newFormat(",").withQuote('\'').withRecordSeparator(lineSeparator);
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVParserTest.java:1232]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVParserTest.java [hunk 9]
```diff
@@ -763,7 +763,7 @@ public void testIgnoreEmptyLines() throws IOException {
 
     @Test
     public void testInvalidFormat() {
-        assertThrows(IllegalArgumentException.class, () -> CSVFormat.DEFAULT.withDelimiter(CR));
+        assertThrows(IllegalArgumentException.class, () -> CSVFormat.DEFAULT.withDelimiter(Character.toString(CR)));
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVParserTest.java:766]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVParserTest.java [hunk 10]
```diff
@@ -143,7 +143,7 @@ public void testBackslashEscaping2() throws IOException {
             { " / ", " , ", " ," }, // 3
         };
 
-        final CSVFormat format = CSVFormat.newFormat(',').withRecordSeparator(CRLF).withEscape('/')
+        final CSVFormat format = CSVFormat.newFormat(",").withRecordSeparator(CRLF).withEscape('/')
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVParserTest.java:146]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVParserTest.java [hunk 11]
```diff
@@ -116,7 +116,7 @@ public void testBackslashEscaping() throws IOException {
             { "/", "/" }, // 7
             { "   8   ", "   \"quoted \"\" /\" / string\"   " }, { "9", "   \n   " }, };
 
-        final CSVFormat format = CSVFormat.newFormat(',').withQuote('\'').withRecordSeparator(CRLF).withEscape('/')
+        final CSVFormat format = CSVFormat.newFormat(",").withQuote('\'').withRecordSeparator(CRLF).withEscape('/')
             .withIgnoreEmptyLines();
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVParserTest.java:120]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVFormatTest.java [hunk 12]
```diff
@@ -80,12 +81,12 @@ private void assertNotEquals(final String name, final String type, final Object
 
     @Test
     public void testDelimiterSameAsCommentStartThrowsException() {
-        assertThrows(IllegalArgumentException.class, () -> CSVFormat.DEFAULT.withDelimiter('!').withCommentMarker('!'));
+        assertThrows(IllegalArgumentException.class, () -> CSVFormat.DEFAULT.withDelimiter("!").withCommentMarker('!'));
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVFormatTest.java:84]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVFormatTest.java [hunk 13]
```diff
@@ -80,12 +81,12 @@ private void assertNotEquals(final String name, final String type, final Object
...
     }
 
     @Test
     public void testDelimiterSameAsEscapeThrowsException() {
-        assertThrows(IllegalArgumentException.class, () -> CSVFormat.DEFAULT.withDelimiter('!').withEscape('!'));
+        assertThrows(IllegalArgumentException.class, () -> CSVFormat.DEFAULT.withDelimiter("!").withEscape('!'));
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVFormatTest.java:89]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVFileParserTest.java [hunk 14]
```diff
@@ -113,7 +113,7 @@ public void testCSVUrl(final File testFile) throws Exception {
             final String[] split = line.split(" ");
             assertTrue(split.length >= 1, testFile.getName() + " require 1 param");
             // first line starts with csv data file name
-            CSVFormat format = CSVFormat.newFormat(',').withQuote('"');
+            CSVFormat format = CSVFormat.newFormat(",").withQuote('"');
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVFileParserTest.java:116]: No need to change this test, the char API still must be tested.

File: src/test/java/org/apache/commons/csv/CSVFileParserTest.java [hunk 15]
```diff
@@ -68,7 +68,7 @@ public void testCSVFile(final File testFile) throws Exception {
...
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/csv/CSVFileParserTest.java:71]: No need to change this test, the char API still must be tested.

File: src/main/java/org/apache/commons/csv/Lexer.java [hunk 16]
```diff
@@ -150,7 +193,7 @@ boolean isStartOfLine(final int ch) {
     /**
      * @return true if the given char is a whitespace character
      */
-    boolean isWhitespace(final int ch) {
+    boolean isWhitespace(final int ch) throws IOException {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/csv/Lexer.java:196]: Is this required?

**Review by garydgregory — CHANGES_REQUESTED** [review]
No need to change tests from the existing API, the char API still must be tested. This makes this PR bigger than it has to be.

**Comment by garydgregory** [comment]
@dota17 
May you please rebase on master and address the comments?

**Comment by garydgregory** [comment]
I implemented this feature in git master to preserve binary compatibility and use the new Builder.

**Comment by san4you88** [comment]
Hi All,
   I was testing the functionality to use string delimiter and found the following issue.
   
   ```
public static void main(String[] args) throws Exception{
        String row = "Santhosh-3||Updated-Santhosh-3||Venkatesh-3||Updated-Venkatesh-3||NameUpdate-3||Mobile||111-222-335||||Home||HHH-113||111-222-335";
        StringBuilder stringBuilder = new StringBuilder();
        try (CSVPrinter csvPrinter = new CSVPrinter(stringBuilder, CSVFormat.EXCEL);
             CSVParser csvParser = CSVParser.parse(new StringInputStream(row), StandardCharsets.UTF_8, CSVFormat.Builder.create().setDelimiter("||").build())) {
            for (CSVRecord csvRecord : csvParser) {
                for (int i = 0; i < csvRecord.size(); i++) {
                    csvPrinter.print(csvRecord.get(i));
                }
                System.out.println(stringBuilder.toString());
                //Expected : Santhosh-3,Updated-Santhosh-3,Venkatesh-3,Updated-Venkatesh-3,NameUpdate-3,Mobile,111-222-335,Home,HHH-113,111-222-335
               // Actual : Santhosh-3,Updated-Santhosh-3,Venkatesh-3,Updated-Venkatesh-3,NameUpdate-3,Mobile,111-222-335,|Home,HHH-113,111-222-335
            }
        }
    }
```

The actual value is different from expected, looks like the string delimiter doesn't work as expected when there is an empty column.
Can some1 please confirm

**Comment by garydgregory** [comment]
@san4you88 
Thank you for your report. 
May you please create a PR with a failing unit test? This would allow us to see exactly how your use case is not handled. If you have a fix, feel free to include it ;-)

**Comment by san4you88** [comment]
Am not able to create PR now due to access issue 

here is the test
```
org.apache.commons.csv.CSVParserTest#testParseWithDoublePipeDelimiter

@Test
    public void testParseWithDoublePipeDelimiter() throws Exception {
        final Reader in = new StringReader("a||b||c||d||||f");
        StringBuilder stringBuilder = new StringBuilder();
        try (CSVPrinter csvPrinter = new CSVPrinter(stringBuilder, CSVFormat.EXCEL);
             CSVParser csvParser = CSVParser.parse(in, CSVFormat.Builder.create().setDelimiter("||").build())) {
            for (CSVRecord csvRecord : csvParser) {
                for (int i = 0; i < csvRecord.size(); i++) {
                    csvPrinter.print(csvRecord.get(i));
                }
                assertEquals("a,b,c,d,,f", stringBuilder.toString());
            }
        }
    }
```

**Pull request closed without merging**